### ISSUE TITLE
IParseNodeFactory.GetRootParseNode should be async

### ISSUE THREAD

**Issue by MihaMarkic** [microsoft/kiota-abstractions-dotnet#223]
Looking at serialization, I see that `IParseNodeFactory.GetRootParseNode` is synchronous even though it reads from a stream.
If nothing else, this is a problem on server side where async read is highly recommended.
Is there a specific reason not being async?

**Comment by MihaMarkic** [comment]
Hi @baywet again,

Yes, getting one of those exceptions and rightly so. No need for stack trace I assume, here is my code where Callback is a generated model (dealing with Webhooks so I have to deserialize it by myself).

```csharp
var pnf = ParseNodeFactoryRegistry.DefaultInstance;
var root = pnf.GetRootParseNode("application/json", content);
var callback = root.GetObjectValue<Callback>(Callback.CreateFromDiscriminatorValue);
```

I know I could force sync write or something else, but the proper way is still doing it in an async way which would be benefical everywhere. Hopefully it will be switched to async model. Also it might not impact much since all(?) the calls are generated and a re-generation to update the code should do the trick.

**Comment by baywet** [comment]
Thanks for the additional information. This is most likely caused by [this line](https://github.com/microsoft/kiota-serialization-json-dotnet/blob/d3af1f6323032dba0c35976e0c0b12e38a49178b/src/JsonParseNodeFactory.cs#L56) which performs a sync stream read under the hood.
The challenge with this change is that adding a new method to an interface is a source breaking change. We could get around it by using [default interface methods](https://www.codejourney.net/csharp-8-default-interface-methods/) but that feature doesn't seem to be supported for dotnet framework (which kiota currently supports).

Quid: what if we updated the json implementation to call ParseAsync(...).GetAwaiter().GetResult() as a workaround? it'd not solve the perf overhead, but at least you should not be getting an exception from the aspnet runtime anymore.

**Comment by MihaMarkic** [comment]
While I understand your concern about breaking changes, those shouldn't affect public surface, right? I mean it's almost everything internal. But if that's a concern, why don't you go using steps - first create required async methods (and perhaps a new interface) in parallel with existing ones. Then switch generated code to async code and mark existing ones as obsolete. Something like that.

AFAIK ParseAsync(...).GetAwaiter().GetResult() is usually a very bad idea because it might deadlock your app - specially client ones, less so server ones where there is no UI context. Also still blocks the thread.

**Comment by baywet** [comment]
We could imagine a situation where we'd have:
- IParseNodeFactory (obsolete attribute on the GetRootParseNode method)
- IAsyncParseNodeFactory (new GetRootParseNodeAsync method, derives from IParseNodeFactory, TODO comment to merge that interface into the parent one with the next major bump)
- all implementations would now implement IAsyncParseNodeFactory, and provide implementations for BOTH methods to avoid breaking compatibility at runtime
- our request adapter [to add a type assertion with IAsyncParseNodeFactory](https://github.com/microsoft/kiota-http-dotnet/blob/6397579b16ba841048a3263a710f6c282c8a1c53/src/HttpClientRequestAdapter.cs#L445) and call the async method instead if the assertion succeeds
- abstractions [helper methods](https://github.com/microsoft/kiota-abstractions-dotnet/blob/3a92aa8bd976001887c92e2c51804d80502bd65f/src/serialization/KiotaSerializer.Deserialization.cs#L52) async overloads (with the type assertion, marking the existing as obsolete)

What do you think?

Is this something you'd be interested in submitting the multiple pull requests for?
Also CC @andrueastman for input

**Comment by andrueastman** [comment]
Should all the sync overloads be marked obsolete? I notice in System.Text.Json for example, the sync overloads are not obsolete. Maybe this should be limited to the overloads taking in a stream?

> We could imagine a situation where we'd have:
> 
> * IParseNodeFactory (obsolete attribute on the GetRootParseNode method)
> * IAsyncParseNodeFactory (new GetRootParseNodeAsync method, derives from IParseNodeFactory, TODO comment to merge that interface into the parent one with the next major bump)
> * all implementations would now implement IAsyncParseNodeFactory, and provide implementations for BOTH methods to avoid breaking compatibility at runtime
> * our request adapter [to add a type assertion with IAsyncParseNodeFactory](https://github.com/microsoft/kiota-http-dotnet/blob/6397579b16ba841048a3263a710f6c282c8a1c53/src/HttpClientRequestAdapter.cs#L445) and call the async method instead if the assertion succeeds
> * abstractions [helper methods](https://github.com/microsoft/kiota-abstractions-dotnet/blob/3a92aa8bd976001887c92e2c51804d80502bd65f/src/serialization/KiotaSerializer.Deserialization.cs#L52) async overloads (with the type assertion, marking the existing as obsolete)
> 
> What do you think?
> 
> Is this something you'd be interested in submitting the multiple pull requests for? Also CC @andrueastman for input

**Comment by MihaMarkic** [comment]
@andrueastman AFAIK everything ends up working on stream in `IParseNodeFactory.IParseNode GetRootParseNode(string contentType, Stream content)`. Perhaps suppress obsolete warning (which results in error) on those public methods taking a non stream content?

**Comment by baywet** [comment]
I'd argue we should mark them as obsolete for the following reasons:
- they were an oversight to begin with.
- we want to clarify to people we'll remove them in the next major release.
- we want to clarify to people which method they should be implementing/using.
- we want to avoid maintaining multiple implementations over time.

**Comment by andrueastman** [comment]
@MihaMarkic I agree. At the end of the day we end working on a stream. So let's make them all obsolete.

**Comment by MihaMarkic** [comment]
@andrueastman @baywet And all of them, by popular demand, are marked as obsolete.

**Comment by andrueastman** [comment]
Closing this one for now as this was resolved via https://github.com/microsoft/kiota-dotnet/pull/225/
